c/uca/sifter/unisifeggs.cpp

```diff
@@ -0,0 +1,440 @@
+    return 0xF8F0 <= c && c <= 0xF8FF;
+}
+
+static bool isSpecialCJKIdeograph(char32_t c) {
+    return 0xFA0E <= c && c <= 0xFA29 &&
+           (c <= 0xFA0F || c == 0xFA11 || c == 0xFA13 || c == 0xFA14 ||
+            c == 0xFA1F || c == 0xFA21 || c == 0xFA23 || c == 0xFA24 ||
+            0xFA27 <= c);
+}
+
+extern "C" int unisift_IsAlphabetic(UInt32 c) {
+    return ucd->BinaryPropertySet("Alphabetic").contains(c) &&
+           !isSpecialCJKIdeograph(c) &&
+           !(c == 0x1E13C || c == 0x1E13D || c == 0x16B42 || c == 0x16B43);
+}
+extern "C" int unisift_IsNonSpacing(UInt32 c) {
+    return ucd->GeneralCategorySet("Mn").contains(c) ||
+           isSifterNonSpacingCombining(c);
+}
+extern "C" int unisift_IsCombining(UInt32 c) {
+    return ucd->GeneralCategorySet("M").contains(c) ||
+           isSifterNonSpacingCombining(c);
+}
+extern "C" int unisift_IsExtender(UInt32 c) {
+    return ucd->BinaryPropertySet("Extender").contains(c);
+}
+extern "C" int unisift_IsCurrency(UInt32 c) {
+    return ucd->GeneralCategorySet("Sc").contains(c);
+}
+extern "C" int unisift_IsPunctuation(UInt32 c) {
+    return ucd->GeneralCategorySet("P").contains(c);
+}
+extern "C" int unisift_IsDiacritic(UInt32 c) {
+    return ucd->BinaryPropertySet("Diacritic").contains(c);
+}
+extern "C" int unisift_IsNumeric(UInt32 c) {
+    return ucd->GeneralCategorySet("N").contains(c);
+}
+extern "C" int unisift_IsDecimalDigit(UInt32 c) {
+    return ucd->GeneralCategorySet("Nd").contains(c);
+}
+extern "C" int unisift_IsWhiteSpace(UInt32 c) {
+    return ucd->BinaryPropertySet("White_Space").contains(c);
+}
+extern "C" int unisift_IsMath(UInt32 c) {
+    return ucd->BinaryPropertySet("Math").contains(c);
+}
+extern "C" int unisift_IsIdeographic(UInt32 c) {
+    return ucd->BinaryPropertySet("Ideographic").contains(c);
+}
+extern "C" int unisift_IsMiscSymbolic(UInt32 c) {
+    return ucd->GeneralCategorySet("Sk").contains(c) ||
+           ucd->GeneralCategorySet("So").contains(c);
+}
+extern "C" int unisift_IsIgnorable(UInt32 c) {
+    return ((ucd->GeneralCategorySet("Cc").contains(c) ||
+             ucd->GeneralCategorySet("Cf").contains(c)) &&
+            !ucd->BinaryPropertySet("White_Space").contains(c)) ||
+           ucd->BinaryPropertySet("Variation_Selector").contains(c) ||
+           c == U'\u00AD';
+}
+
+extern "C" int unisift_IsUpper(UInt32 c) {
+    return ucd->BinaryPropertySet("Uppercase").contains(c);
+}
+
+extern "C" UInt32 unisift_ToLower(UInt32 c) {
+    return ucd->SimpleLowercaseMapping(c);
+}
+
+extern "C" UInt32 unisift_ToUpper(UInt32 c) {
+    return ucd->SimpleUppercaseMapping(c);
+}
+
+extern "C" int unisift_ToIntValue(UInt32 c) {
+    // TODO(egg): The conversion to the return type is UB for large numbers.
+    return ucd->NaturalNumericValue(c).value_or(
+        ucd->CodePointsWithNumericValue().contains(c) ? -2 : -1);
+}
+
+extern "C" int unisift_GetCombiningClass(UInt32 c) {
+    return ucd->CanonicalCombiningClass(c);
+}
```

Comment:
Don't you also need unisift_unistrcpy() and unisift_unistrcat()?
They were in Ken's version of unisifex.c, and I assumed that the sifter code calls them.

Reply:
Apparently it does not.